copy/multiple.go

```diff
@@ -27,23 +32,90 @@ const (
+	// is `instanceCopyClone`
+	compressionVariant            *compression.Algorithm
+	cloneAddPlatform              *imgspecv1.Platform
+	cloneAddAnnotations           map[string]string
+	cloneAddCompressionAlgorithms []compression.Algorithm
+}
+
+// internal type only to make imgspecv1.Platform comparable
+type platformComparable struct {
+	Architecture string
+	OS           string
+	OSVersion    string
+	OSFeatures   string
+	Variant      string
+}
+
+// Converts imgspecv1.Platform to a comparable format.
+func platformV1ToPlatformComparable(platform imgspecv1.Platform) platformComparable {
+	return platformComparable{Architecture: platform.Architecture,
+		OS:         platform.OS,
+		OSFeatures: strings.Join(platform.OSFeatures, ","),
+		OSVersion:  platform.OSVersion,
+		Variant:    platform.Variant,
+	}
+}
+
+// platformCompresionPair prepares a mapping of platformComparable -> CompressionAlgorithmNames for given digests
+func platformCompresionPair(list internalManifest.List, instanceDigests []digest.Digest) (map[platformComparable]*set.Set[string], error) {
+	platformCompressionMap := make(map[platformComparable]*set.Set[string])
+	for _, instanceDigest := range instanceDigests {
+		instanceDetails, err := list.Instance(instanceDigest)
+		if err != nil {
+			return nil, fmt.Errorf("getting details for instance %s: %w", instanceDigest, err)
+		}
+		platform := platformV1ToPlatformComparable(*instanceDetails.ReadOnly.Platform)
+		_, valueExists := platformCompressionMap[platform]
+		if !valueExists {
+			platformCompressionMap[platform] = set.New[string]()
+		}
+		set := platformCompressionMap[platform]
+		set.AddSlice(instanceDetails.ReadOnly.CompressionAlgorithmNames)
+	}
+	return platformCompressionMap, nil
 }
 
 // prepareInstanceCopies prepares a list of instances which needs to copied to the manifest list.
-func prepareInstanceCopies(instanceDigests []digest.Digest, options *Options) []instanceCopy {
+func prepareInstanceCopies(list internalManifest.List, instanceDigests []digest.Digest, options *Options) ([]instanceCopy, error) {
 	res := []instanceCopy{}
+	platformCompressionMap, err := platformCompresionPair(list, instanceDigests)
+	if err != nil {
+		return res, err
+	}
 	for i, instanceDigest := range instanceDigests {
 		if options.ImageListSelection == CopySpecificImages &&
 			!slices.Contains(options.Instances, instanceDigest) {
 			logrus.Debugf("Skipping instance %s (%d/%d)", instanceDigest, i+1, len(instanceDigests))
 			continue
 		}
+		instanceDetails, err := list.Instance(instanceDigest)
+		if err != nil {
+			return res, fmt.Errorf("getting details for instance %s: %w", instanceDigest, err)
+		}
+		for _, compressionVariant := range options.EnsureCompressionVariantsExist {
+			platform := platformV1ToPlatformComparable(*instanceDetails.ReadOnly.Platform)
+			compressionList := platformCompressionMap[platform]
+			if !slices.Contains(compressionList.Values(), compressionVariant.Algorithm.Name()) {
```

Comment:
A design question here is whether we should be looking for any match that contains a Zstd layer, or only for exclusively-Zstd images.

OTOH, right now we can’t tell the difference, so it doesn‘t matter anyway… simpler code is immediately valuable.